src/NServiceBus.Core.Analyzer.Tests/Helpers/DiagnosticVerifier.cs

```diff
@@ -0,0 +1,221 @@
+            var systemTasks = MetadataReference.CreateFromFile(System.Reflection.Assembly.Load("System.Threading.Tasks").Location);
+            var systemRuntime = MetadataReference.CreateFromFile(System.Reflection.Assembly.Load("System.Runtime").Location);
+            solution = solution.AddMetadataReferences(projectId, new[]
+            {
+                netstandard,
+                systemRuntime,
+                systemTasks
+            });
+#endif
+
+            var documentIndex = 0;
+            foreach (var source in sources)
+            {
+                var fileName = "Test" + documentIndex + ".cs";
+                solution = solution.AddDocument(DocumentId.CreateNewId(projectId, fileName), fileName, SourceText.From(source));
+                documentIndex++;
+            }
+
+            var documents = solution.GetProject(projectId).Documents.ToList();
+
+            if (sources.Length != documents.Count)
+            {
+                throw new InvalidOperationException("The number of documents created does not match the number of sources.");
+            }
+
+            var diagnostics = new List<Diagnostic>();
+            foreach (var project in new HashSet<Project>(documents.Select(document => document.Project)))
+            {
+                var compilationWithAnalyzers = (await project.GetCompilationAsync()).WithAnalyzers(ImmutableArray.Create(analyzer));
+
+                foreach (var diagnostic in await compilationWithAnalyzers.GetAnalyzerDiagnosticsAsync())
+                {
+                    if (diagnostic.Location == Location.None || diagnostic.Location.IsInMetadata)
+                    {
+                        diagnostics.Add(diagnostic);
+                    }
+                    else
+                    {
+                        foreach (var document in documents)
+                        {
+                            if (await document.GetSyntaxTreeAsync() == diagnostic.Location.SourceTree)
+                            {
+                                diagnostics.Add(diagnostic);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return diagnostics.OrderBy(diagnostic => diagnostic.Location.SourceSpan.Start).ToArray();
+        }
+
+        static string FormatDiagnostics(DiagnosticAnalyzer analyzer, params Diagnostic[] diagnostics)
+        {
+            var builder = new StringBuilder();
+            for (var diagnosticIndex = 0; diagnosticIndex < diagnostics.Length; ++diagnosticIndex)
+            {
+                builder.AppendLine("// " + diagnostics[diagnosticIndex]);
+
+                var analyzerType = analyzer.GetType();
+                var descriptors = analyzer.SupportedDiagnostics;
+
+                foreach (var descriptor in descriptors
+                    .Where(descriptor => descriptor?.Id == diagnostics[diagnosticIndex].Id))
```

Comment:
Another place where I'd prefer to have this on a single line.

Reply:
👎 😉